tests/common.c

```diff
@@ -0,0 +1,307 @@
+const char *
+ws_strerror(int err)
+{
+  static char ws_err[50];
+
+  if(err >= WSABASEERR) {
+    snprintf(ws_err, sizeof(ws_err), "Winsock err: %d", err);
+    return ws_err;
+  }
+  /* Assume a CRT error */
+  return (strerror)(err);
+}
+#endif
+
+/*
+ * Write n bytes from buf to the provided fd (on Windows, this must be
+ * stdout/stderr or a file, not a socket), retrying short writes until
+ * we finish or hit an error. Assumes fd is blocking and therefore doesn't
+ * handle EAGAIN. Returns 0 for success or 1 for error.
+ */
+int
+write_all(int fd, const char *buf, int n)
+{
+  int m = 0;
+
+  while(n > 0) {
+    m = write(fd, buf, n);
+    if(m < 0) {
+      perror("write_all");
+      return 1;
+    }
+    n -= m;
+  }
+  return 0;
+}
+
+/*
+ * Set a socket to be nonblocking.
+ *
+ * Returns CRUSTLS_DEMO_OK on success, CRUSTLS_DEMO_ERROR on error.
+ */
+enum crustls_demo_result
+nonblock(int sockfd)
+{
+#ifdef _WIN32
+  u_long nonblock = 1UL;
+
+  if(ioctlsocket(sockfd, FIONBIO, &nonblock) != 0) {
+    perror("Error setting socket nonblocking");
+    return CRUSTLS_DEMO_ERROR;
+  }
+#else
+  int flags;
+  flags = fcntl(sockfd, F_GETFL, 0);
+  if(flags < 0) {
+    perror("getting socket flags");
+    return CRUSTLS_DEMO_ERROR;
+  }
+  flags = fcntl(sockfd, F_SETFL, flags | O_NONBLOCK);
+  if(flags < 0) {
+    perror("setting socket nonblocking");
+    return CRUSTLS_DEMO_ERROR;
+  }
+#endif
+  return CRUSTLS_DEMO_OK;
+}
+
+int
+read_cb(void *userdata, unsigned char *buf, size_t len, size_t *out_n)
+{
+  ssize_t n = 0;
+  struct conndata_t *conn = (struct conndata_t *)userdata;
+  n = recv(conn->fd, buf, len, 0);
+  if(n < 0) {
+    return errno;
+  }
+  if(out_n != NULL) {
+    *out_n = n;
+  }
+  return 0;
+}
+
+int
+write_cb(void *userdata, const unsigned char *buf, size_t len, size_t *out_n)
+{
+  ssize_t n = 0;
+  struct conndata_t *conn = (struct conndata_t *)userdata;
+
+  n = send(conn->fd, buf, len, 0);
+  if(n < 0) {
+    return errno;
+  }
+  if(out_n != NULL) {
+    *out_n = n;
+  }
+  return 0;
+}
+
+size_t
+bytevec_available(struct bytevec *vec)
+{
+  return vec->capacity - vec->len;
+}
+
+char *
+bytevec_writeable(struct bytevec *vec)
+{
+  return vec->data + vec->len;
+}
+
+void
+bytevec_consume(struct bytevec *vec, size_t n)
+{
+  vec->len += n;
+}
+
+// Ensure there are at least n bytes available between vec->len and
+// vec->capacity. If this requires reallocating, this may return
+// CRUSTLS_DEMO_ERROR.
+enum crustls_demo_result
+bytevec_ensure_available(struct bytevec *vec, size_t n)
+{
+  size_t available = vec->capacity - vec->len;
+  size_t newsize;
+  void *newdata;
+  if(available < n) {
+    newsize = vec->len + n;
+    if(newsize < vec->capacity * 2) {
+      newsize = vec->capacity * 2;
+    }
+    newdata = realloc(vec->data, newsize);
+    if(newdata == NULL) {
+      fprintf(stderr, "out of memory trying to get %ld bytes\n", newsize);
+      return CRUSTLS_DEMO_ERROR;
+    }
+    vec->data = (char *)newdata;
```

Comment:
nit: Is this cast necessary? C will let you assign the `void *` returned from `calloc(3)` to `char *` without a warning. The cast is only needed if you also want to compile this as C++, where the compilers are pickier, but then you'd also have to conditionally put in `extern "C" {}` around everything.